tesseract_command_language/src/utils/utils.cpp

```diff
@@ -325,6 +330,137 @@ void generateSkeletonSeedHelper(CompositeInstruction& composite_instructions)
   }
 }
 
+bool toDelimitedFile(const CompositeInstruction& composite_instructions, const std::string& file_path, char separator)
+{
+  static const Eigen::IOFormat eigen_format(
+      Eigen::StreamPrecision, Eigen::DontAlignCols, " ", std::string(&separator, 1));
+  std::ofstream myfile;
+  myfile.open(file_path);
+
+  std::vector<std::reference_wrapper<const Instruction>> mi = flatten(composite_instructions, &moveFilter);
+
+  // Write Joint names as header
+  std::vector<std::string> joint_names = getJointNames(mi.front().get().cast_const<MoveInstruction>()->getWaypoint());
+  std::copy(joint_names.begin(), joint_names.end(), std::ostream_iterator<std::string>(myfile, &separator));
+  myfile << "\n";
+
+  // Write Positions
+  for (const auto& i : mi)
+  {
+    Eigen::VectorXd p = getJointPosition(i.get().cast_const<MoveInstruction>()->getWaypoint());
+    std::stringstream xyz_string;
+    myfile << p.format(eigen_format) << "\n";
+  }
+
+  myfile.close();
+  return true;
+}
+
+bool operator==(const Waypoint& lhs, const Waypoint& rhs)
+{
+  if (lhs.getType() == rhs.getType())
+  {
+    if (lhs.getType() == static_cast<int>(WaypointType::NULL_WAYPOINT))
+      return true;
+    else if (lhs.getType() == static_cast<int>(WaypointType::JOINT_WAYPOINT))
+      return ((*lhs.cast_const<JointWaypoint>()) == (*rhs.cast_const<JointWaypoint>()));
+    else if (lhs.getType() == static_cast<int>(WaypointType::CARTESIAN_WAYPOINT))
+      return ((*lhs.cast_const<CartesianWaypoint>()) == (*rhs.cast_const<CartesianWaypoint>()));
+    else if (lhs.getType() == static_cast<int>(WaypointType::STATE_WAYPOINT))
+      return ((*lhs.cast_const<StateWaypoint>()) == (*rhs.cast_const<StateWaypoint>()));
+    else
+      return false;
+  }
+
+  return false;
+}
+
+bool operator!=(const Waypoint& lhs, const Waypoint& rhs) { return !operator==(lhs, rhs); }
```

Comment:
I guess it might make sense for the Waypoint and Instruction ones to be here. But should MoveInstruction operators, for example, go in the class?

Reply:
Everything except Waypoint and Instruction compare operator is located in the associated classes.